src/Symfony/Component/Security/Http/Authentication/AuthenticatorManager.php

```diff
@@ -0,0 +1,242 @@
+        $request->attributes->remove('_security_authenticators');
+        if (!$authenticators) {
+            return null;
+        }
+
+        return $this->executeAuthenticators($authenticators, $request);
+    }
+
+    /**
+     * @param AuthenticatorInterface[] $authenticators
+     */
+    private function executeAuthenticators(array $authenticators, Request $request): ?Response
+    {
+        foreach ($authenticators as $authenticator) {
+            // recheck if the authenticator still supports the listener. supports() is called
+            // eagerly (before token storage is initialized), whereas authenticate() is called
+            // lazily (after initialization). This is important for e.g. the AnonymousAuthenticator
+            // as its support is relying on the (initialized) token in the TokenStorage.
+            if (false === $authenticator->supports($request)) {
+                if (null !== $this->logger) {
+                    $this->logger->debug('Skipping the "{authenticator}" authenticator as it did not support the request.', ['authenticator' => \get_class($authenticator)]);
+                }
+
+                continue;
+            }
+
+            $response = $this->executeAuthenticator($authenticator, $request);
+            if (null !== $response) {
+                if (null !== $this->logger) {
+                    $this->logger->debug('The "{authenticator}" authenticator set the response. Any later authenticator will not be called', ['authenticator' => \get_class($authenticator)]);
+                }
+
+                return $response;
+            }
+        }
+
+        return null;
+    }
+
+    private function executeAuthenticator(AuthenticatorInterface $authenticator, Request $request): ?Response
+    {
+        try {
+            // get the passport from the Authenticator
+            $passport = $authenticator->authenticate($request);
+
+            // check the passport (e.g. password checking)
+            $event = new VerifyAuthenticatorCredentialsEvent($authenticator, $passport);
+            $this->eventDispatcher->dispatch($event);
+
+            // check if all badges are resolved
+            $passport->checkIfCompletelyResolved();
+
+            // create the authenticated token
+            $authenticatedToken = $authenticator->createAuthenticatedToken($passport, $this->firewallName);
+            if (true === $this->eraseCredentials) {
+                $authenticatedToken->eraseCredentials();
+            }
+
+            if (null !== $this->eventDispatcher) {
+                $this->eventDispatcher->dispatch(new AuthenticationSuccessEvent($authenticatedToken), AuthenticationEvents::AUTHENTICATION_SUCCESS);
+            }
+
+            if (null !== $this->logger) {
+                $this->logger->info('Authenticator successful!', ['token' => $authenticatedToken, 'authenticator' => \get_class($authenticator)]);
```

Comment:
i would put the key authenticator first
and create a var for \get_class($authenticator) as you use it below as well :)

Reply:
As get_class is an opcode optimized function, I would guess that calling it twice is better than saving it in a variable?

Reply:
was just asking :)